autoscheduler/frontend/src/redux/actions/courseCards.ts

```diff
@@ -93,6 +93,45 @@ export function parseSections(arr: any[]): SectionSelected[] {
   return sections;
 }
 
+function sortProfessorGroupsBySectionNumber(sections: SectionSelected[]): SectionSelected[] {
+  // return if there's nothing to sort
+  if (sections.length < 2) return sections;
+
+  interface ProfInfo {
+    name: string;
+    sectionNum: string;
+    start: number;
+    end: number;
+  }
+  const getProfInfo = (idx: number): ProfInfo => ({
+    name: sections[idx].section.instructor.name,
+    sectionNum: sections[idx].section.sectionNum,
+    start: idx,
+    end: null as number,
+  });
+  let profs = [getProfInfo(0)];
+  for (let i = 0; i < sections.length; i++) {
+    const currProf = getProfInfo(i);
+    if (profs[profs.length - 1].name !== currProf.name) {
+      // save end index for last professor
+      profs[profs.length - 1].end = i;
+      // and start index for new professor
+      profs.push(getProfInfo(i));
+    }
+  }
+  // set end index for last professor group
+  profs[profs.length - 1].end = sections.length;
+
+  // sort professor groups by section number
+  profs = profs.sort((a, b) => a.sectionNum.localeCompare(b.sectionNum));
+
+  // rebuild sections
+  const sorted: SectionSelected[] = [];
+  profs.forEach((prof) => sorted.push(...sections.slice(prof.start, prof.end)));
+
+  return sorted;
+}
+
```

Comment:
A couple things about this sort:
- I don't think it works exactly as intended, for example when I try to see the results for MATH 151 in 202031, it shows TBA honors, then TBA non-honors, then more TBA honors, then more TBA non-honors sections
- I don't really like how much custom code this uses to sort, there are a lot of points it could potentially make mistakes and it isn't very easy to read

I've tested this code and it seems to work fine (it also makes it so we don't need to separately sort by instructor names like we do on line 155), another advantage is that it's easy to adapt if we want to change the sort order (for example, have `sectionToArr` check if the prof name is TBA, and if so set the field to ZZZZ or something so it shows up last):
```js
function sortSections(sections: SectionSelected[]): SectionSelected[] {
  // converts a section to an array of relevant fields for sort
  const sectionToArr = (section: SectionSelected, idx: number): any[] => (
    [section.section.instructor.name, section.section.sectionNum, idx]);
  const sortedArrs = sections.map(sectionToArr).sort();
  // sort sections based on their positions in the sorted array of arrays
  const sorted = sortedArrs.map((section) => sections[section[2]]);
  return sorted;
}
```

Reply:
Wow, that is a really clean way to write it! Thanks for the suggestion. I'll add a test for the TBA thing, then rewrite the sort using this version

Reply:
I came up with a way to do it that displays professors ordered by their lowest section number (which I believe is how we wanted to do it) but is a little more complicated:
```js
function sortSections(sections: SectionSelected[]): SectionSelected[] {
  // sort sections by sectionNum
  const sorted = sections.sort((a: SectionSelected, b: SectionSelected) => {
    if (a.section.sectionNum < b.section.sectionNum) return -1;
    return 1;
  });
  const sectionsForProfs: Map<string, SectionSelected[]> = new Map();
  // maps maintain key insertion order, so add all sections to map and remember order of professors
  sorted.forEach((section) => {
    const instructorName = section.section.instructor.name;
    if (sectionsForProfs.has(instructorName)) sectionsForProfs.get(instructorName).push(section);
    else sectionsForProfs.set(instructorName, [section]);
  });
  // sections are now grouped by professor and sorted by section num
  return [].concat(...sectionsForProfs.values());
}
```
I think the following change correctly puts TBA sections last, but I haven't found any courses that I can test this with/made any test data for it:
```js
function sortSections(sections: SectionSelected[]): SectionSelected[] {
  // sort sections by sectionNum
  const sorted = sections.sort((a: SectionSelected, b: SectionSelected) => {
    if (a.section.sectionNum < b.section.sectionNum) return -1;
    return 1;
  });
  const sectionsForProfs: Map<string, SectionSelected[]> = new Map();
  // maps maintain key insertion order, so add all sections to map and remember order of professors
  const TBASections: SectionSelected[] = [];
  sorted.forEach((section) => {
    const instructorName = section.section.instructor.name;
    if (instructorName === 'TBA') {
      TBASections.push(section);
    } else if (sectionsForProfs.has(instructorName)) {
      sectionsForProfs.get(instructorName).push(section);
    } else sectionsForProfs.set(instructorName, [section]);
  });
  // sections are now grouped by professor and sorted by section num
  return [].concat(...sectionsForProfs.values(), TBASections);
}
```